modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/source/poll/PollingSourceWrapper.java

```diff
@@ -127,13 +137,34 @@ public void onStart(SourceCallback<T, A> sourceCallback) throws MuleException {
     watermarkObjectStore = objectStoreManager.getOrCreateObjectStore(formatKey(WATERMARK_OS_NAME_SUFFIX),
                                                                      unmanagedPersistent());
 
-    executor = schedulerService.customScheduler(SchedulerConfig.config()
+    postReconnectionExecutor = createScheduler("postReconnectionExecutor");
+  }
+
+  private org.mule.runtime.api.scheduler.Scheduler createScheduler(String name) {
+    return schedulerService.customScheduler(SchedulerConfig.config()
         .withMaxConcurrentTasks(1)
         .withWaitAllowed(true)
-        .withName(formatKey("executor")));
+        .withName(formatKey(name)));
+  }
+
+  @Override
+  public void onStart(SourceCallback<T, A> sourceCallback) throws MuleException {
+    sourceCallbackReference.set(sourceCallback);
+
+    delegate.onStart(sourceCallback);
 
     stopRequested.set(false);
-    scheduler.schedule(executor, () -> poll(sourceCallback));
+
+    if (reconnection.get()) {
+      postReconnectionExecutor.schedule(pollAction, 0, TimeUnit.SECONDS);
```

Comment:
why does this happen on a separate thread? This looks like a huge race condition. Can you clarify?